apps/opik-backend/src/main/java/com/comet/opik/domain/SpanService.java

```diff
@@ -244,4 +250,138 @@ public Mono<Boolean> validateSpanWorkspace(@NonNull String workspaceId, @NonNull
         return spanDAO.getSpanWorkspace(spanIds)
                 .map(spanWorkspace -> spanWorkspace.stream().allMatch(span -> workspaceId.equals(span.workspaceId())));
     }
+
+    @Trace(dispatcher = true)
+    public Mono<Void> create(SpanBatch batch) {
+
+        if (batch.spans().isEmpty()) {
+            return Mono.empty();
+        }
+
+        List<String> projectNames = batch.spans()
+                .stream()
+                .map(Span::projectName)
+                .distinct()
+                .toList();
+
+        Mono<List<Span>> resolveProjects = Flux.fromIterable(projectNames)
+                .flatMap(this::resolveProject)
+                .collectList()
+                .map(projects -> bindSpanToProjectAndId(batch, projects))
+                .subscribeOn(Schedulers.boundedElastic());
+
+        return resolveProjects
+                .flatMap(spans -> {
+                    List<UUID> spanIds = spans.stream().map(Span::id).distinct().toList();
+
+                    return lockService.lockAll(spanIds, SPAN_KEY)
+                            .flatMap(locks -> mergeSpans(spans, spanIds)
+                                    .flatMap(spanDAO::batchInsert)
+                                    .doFinally(signalType -> lockService.unlockAll(locks).subscribe())
+                                    .subscribeOn(Schedulers.boundedElastic()))
+                            .then();
+                });
+    }
+
+    private Mono<List<Span>> mergeSpans(List<Span> spans, List<UUID> spanIds) {
+        return spanDAO.getByIds(spanIds)
+                .map(existingSpans -> {
+                    Map<UUID, Span> existingSpanMap = existingSpans.stream()
+                            .collect(Collectors.toMap(Span::id, Function.identity()));
+
+                    return spans.stream()
+                            .collect(Collectors.groupingBy(Span::id))
+                            .entrySet()
+                            .stream()
+                            .map(groupedSpans -> {
+                                // START state resolution from existing spans
+                                Span currentSpan = existingSpanMap.getOrDefault(groupedSpans.getKey(),
+                                        Span.builder().build());
+
+                                return groupedSpans
+                                        .getValue()
+                                        .stream()
+                                        .reduce(currentSpan, this::mergeSpans);
+                            }).toList();
+                });
+    }
+
+    private Span mergeSpans(Span currentSpan, Span receivedSpan) {
```

Comment:
Merging spans should no longer be needed as there's no going to be update flow anymore, specially for batch endpoints.

We should get rid of all the related code to merging spans. Maybe the only reasonable exception would be checking for the existence of the ID. But even that, it's hard in a batch scenario, so most likely should be removed as well.

Reply:
We can definitely remove the merge logic. The only concerns are updates and the creation time. In this sense, the billing will use it for traces and probably spans as well. What do you think?